core/kazoo_number_manager/src/carriers/knm_voxbone.erl

```diff
@@ -0,0 +1,332 @@
+    end.
+
+%%------------------------------------------------------------------------------
+%% @doc get and process available DID quantities
+%% @end
+%%------------------------------------------------------------------------------
+-spec do_search(kz_term:api_pos_integer(), kz_term:proplist()) -> {'error', any()} | map().
+do_search(Quantity, SearchOptions) ->
+    QueryString = knm_voxbone_util:required_params() ++  SearchOptions,
+    case knm_voxbone_util:voxbone_request('get', <<"inventory/didgroup">>, QueryString) of
+        {'error', Reason} -> {'error',  Reason};
+        {'ok', DidGroups} -> process_search_results(Quantity, DidGroups)
+    end.
+
+%%------------------------------------------------------------------------------
+%% @doc reserve quantities from DID groups and build order
+%% @end
+%%------------------------------------------------------------------------------
+-spec process_search_results(kz_term:api_pos_integer(), kz_json:object()) -> map().
+process_search_results(Quantity, Response) ->
+    DidGroups = kz_json:get_value(<<"didGroups">>, Response, kz_json:new()),
+    resolve_order_quantities(DidGroups, Quantity, #{total => 0, order => [], cost => 0}).
+
+%%------------------------------------------------------------------------------
+%% @doc reconcile a quantity of DIDs to purchase from a given DIDGroup
+%% @end
+%%------------------------------------------------------------------------------
+-spec resolve_order_quantities(kz_json:objects(), kz_term:api_pos_integer(), kz_term:proplist()) ->  map().
+resolve_order_quantities([], _Remaining, Acc) ->
+    Acc;
+resolve_order_quantities(_DidGroups, 0=Remaining, Acc) ->
+    resolve_order_quantities([], Remaining, Acc);
+resolve_order_quantities([DidGroup | Rest], Remaining, #{total :=Total, order := Order, cost := Cost}=_Acc) ->
+    DidGroupId = kz_json:get_value(<<"didGroupId">>, DidGroup),
+    Available = kz_json:get_value(<<"stock">>, DidGroup),
+    UnitPrice = kz_json:get_value(<<"setup100">>, DidGroup) + kz_json:get_value(<<"monthly100">>, DidGroup),
+    {Remaining2, Reserve} = maybe_reserve_did_qty(Available, Remaining),
+    NewCost = Reserve * (UnitPrice / 100),
+    lager:debug("ordering ~p of ~p dids from didGroup ~p.", [Reserve, Available, DidGroupId]),
+    resolve_order_quantities(Rest, Remaining2, #{total => Total+Reserve, order => Order ++ [{DidGroupId, Reserve}], cost => NewCost + Cost}).
+
+%%------------------------------------------------------------------------------
+%% @doc determine what amount from available inventory can be reserved
+%% @end
+%%------------------------------------------------------------------------------
+-spec maybe_reserve_did_qty(kz_term:api_pos_integer(), kz_term:api_pos_integer()) -> {kz_term:api_pos_integer(), kz_term:api_pos_integer()}.
+maybe_reserve_did_qty(Available, Need) when Available < Need ->
+    {Need - Available, Available};
+maybe_reserve_did_qty(_Available, Need) ->
+    {0, Need}.
```

Comment:
Similarly, this clause means the function can return `non_neg_integer()`

Reply:
Actually, can it even be 'undefined'? Probably just simplify all this to `non_neg_integer()`